File: xiangqi/xiangqi-core/src/rules.rs
use crate::types::*;
use crate::board::Board;

pub fn is_legal_move(board: &Board, from: Position, to: Position) -> bool {
    if !from.is_valid() || !to.is_valid() {
        return false;
    }
    if from == to {
        return false;
    }

    let piece = board.piece_at(from);
    if piece.is_none() {
        return false;
    }
    let piece = piece.unwrap();
    if piece.side != board.side_to_move() {
        return false;
    }

    let target = board.piece_at(to);
    if let Some(target_piece) = target {
        if target_piece.side == piece.side {
            return false;
        }
    }

    if !is_piece_move_valid(board, piece, from, to) {
        return false;
    }

    let mut temp_board = board.clone();
    temp_board.make_move_internal(Move::new(from, to));
    if is_in_check(&temp_board, piece.side) {
        return false;
    }

    true
}

fn is_piece_move_valid(board: &Board, piece: Piece, from: Position, to: Position) -> bool {
    match piece.piece_type {
        PieceType::King => is_king_move_valid(piece, from, to),
        PieceType::Advisor => is_advisor_move_valid(piece, from, to),
        PieceType::Bishop => is_bishop_move_valid(board, piece, from, to),
        PieceType::Knight => is_knight_move_valid(board, from, to),
        PieceType::Rook => is_rook_move_valid(board, from, to),
        PieceType::Cannon => is_cannon_move_valid(board, from, to),
        PieceType::Pawn => is_pawn_move_valid(piece, from, to),
    }
}

fn is_king_move_valid(piece: Piece, from: Position, to: Position) -> bool {
    if !to.in_palace(piece.side) {
        return false;
    }
    let dx = (to.x as i32 - from.x as i32).abs();
    let dy = (to.y as i32 - from.y as i32).abs();
    dx + dy == 1
}

fn is_advisor_move_valid(piece: Piece, from: Position, to: Position) -> bool {
    if !to.in_palace(piece.side) {
        return false;
    }
    let dx = (to.x as i32 - from.x as i32).abs();
    let dy = (to.y as i32 - from.y as i32).abs();
    dx == 1 && dy == 1
}

fn is_bishop_move_valid(board: &Board, piece: Piece, from: Position, to: Position) -> bool {
    if !to.in_own_territory(piece.side) {
        return false;
    }
    let dx = (to.x as i32 - from.x as i32).abs();
    let dy = (to.y as i32 - from.y as i32).abs();
    if dx != 2 || dy != 2 {
        return false;
    }

    let eye_x = from.x as i32 + (to.x as i32 - from.x as i32) / 2;
    let eye_y = from.y as i32 + (to.y as i32 - from.y as i32) / 2;
    if board.piece_at(Position::new(eye_x as u8, eye_y as u8)).is_some() {
        return false;
    }

    true
}

fn is_knight_move_valid(board: &Board, from: Position, to: Position) -> bool {
    let dx = to.x as i32 - from.x as i32;
    let dy = to.y as i32 - from.y as i32;
    let adx = dx.abs();
    let ady = dy.abs();

    if !((adx == 1 && ady == 2) || (adx == 2 && ady == 1)) {
        return false;
    }

    let (block_x, block_y) = if adx == 2 {
        (from.x as i32 + dx / 2, from.y as i32)
    } else {
        (from.x as i32, from.y as i32 + dy / 2)
    };

    if board.piece_at(Position::new(block_x as u8, block_y as u8)).is_some() {
        return false;
    }

    true
}

fn is_rook_move_valid(board: &Board, from: Position, to: Position) -> bool {
    if from.x != to.x && from.y != to.y {
        return false;
    }
    is_path_clear(board, from, to)
}

fn is_cannon_move_valid(board: &Board, from: Position, to: Position) -> bool {
    if from.x != to.x && from.y != to.y {
        return false;
    }

    let target = board.piece_at(to);
    if target.is_some() {
        let count = count_pieces_between(board, from, to);
        count == 1
    } else {
        is_path_clear(board, from, to)
    }
}

fn is_pawn_move_valid(piece: Piece, from: Position, to: Position) -> bool {
    let dx = (to.x as i32 - from.x as i32).abs();
    let dy = to.y as i32 - from.y as i32;

    let forward = match piece.side {
        Side::Red => -1,
        Side::Black => 1,
    };

    let past_river = from.is_past_river(piece.side);

    if past_river {
        if dy == forward && dx == 0 {
            return true;
        }
        if dy == 0 && dx == 1 {
            return true;
        }
        false
    } else {
        dy == forward && dx == 0
    }
}

fn is_path_clear(board: &Board, from: Position, to: Position) -> bool {
    let dx = to.x as i32 - from.x as i32;
    let dy = to.y as i32 - from.y as i32;
    let steps = dx.abs().max(dy.abs());
    if steps == 0 {
        return false;
    }
    let step_x = dx / steps;
    let step_y = dy / steps;

    for i in 1..steps {
        let x = from.x as i32 + step_x * i;
        let y = from.y as i32 + step_y * i;
        if board.piece_at(Position::new(x as u8, y as u8)).is_some() {
            return false;
        }
    }
    true
}

fn count_pieces_between(board: &Board, from: Position, to: Position) -> u32 {
    let dx = to.x as i32 - from.x as i32;
    let dy = to.y as i32 - from.y as i32;
    let steps = dx.abs().max(dy.abs());
    let step_x = dx / steps;
    let step_y = dy / steps;

    let mut count = 0;
    for i in 1..steps {
        let x = from.x as i32 + step_x * i;
        let y = from.y as i32 + step_y * i;
        if board.piece_at(Position::new(x as u8, y as u8)).is_some() {
            count += 1;
        }
    }
    count
}

pub fn is_in_check(board: &Board, side: Side) -> bool {
    let king_pos = match side {
        Side::Red => board.red_king_pos(),
        Side::Black => board.black_king_pos(),
    };

    let opponent = side.opponent();
    for pos in board.all_pieces(opponent) {
        let piece = board.piece_at(pos).unwrap();
        if can_attack(board, piece, pos, king_pos) {
            return true;
        }
    }

    if kings_opposing(board) {
        return true;
    }

    false
}

fn can_attack(board: &Board, piece: Piece, from: Position, to: Position) -> bool {
    match piece.piece_type {
        PieceType::King => from.x == to.x && is_path_clear(board, from, to),
        PieceType::Advisor => {
            let dx = (to.x as i32 - from.x as i32).abs();
            let dy = (to.y as i32 - from.y as i32).abs();
            dx == 1 && dy == 1
        }
        PieceType::Bishop => {
            let dx = (to.x as i32 - from.x as i32).abs();
            let dy = (to.y as i32 - from.y as i32).abs();
            if dx == 2 && dy == 2 {
                let eye_x = from.x as i32 + (to.x as i32 - from.x as i32) / 2;
                let eye_y = from.y as i32 + (to.y as i32 - from.y as i32) / 2;
                board.piece_at(Position::new(eye_x as u8, eye_y as u8)).is_none()
            } else {
                false
            }
        }
        PieceType::Knight => {
            let dx = (to.x as i32 - from.x as i32).abs();
            let dy = (to.y as i32 - from.y as i32).abs();
            if (dx == 1 && dy == 2) || (dx == 2 && dy == 1) {
                let (block_x, block_y) = if dx == 2 {
                    (from.x as i32 + (to.x as i32 - from.x as i32) / 2, from.y as i32)
                } else {
                    (from.x as i32, from.y as i32 + (to.y as i32 - from.y as i32) / 2)
                };
                board.piece_at(Position::new(block_x as u8, block_y as u8)).is_none()
            } else {
                false
            }
        }
        PieceType::Rook => (from.x == to.x || from.y == to.y) && is_path_clear(board, from, to),
        PieceType::Cannon => {
            if from.x == to.x || from.y == to.y {
                count_pieces_between(board, from, to) == 1
            } else {
                false
            }
        }
        PieceType::Pawn => {
            let dx = (to.x as i32 - from.x as i32).abs();
            let dy = to.y as i32 - from.y as i32;
            let forward = match piece.side {
                Side::Red => -1,
                Side::Black => 1,
            };
            let past_river = from.is_past_river(piece.side);
            if past_river {
                (dy == forward && dx == 0) || (dy == 0 && dx == 1)
            } else {
                dy == forward && dx == 0
            }
        }
    }
}

fn kings_opposing(board: &Board) -> bool {
    let red_king = board.red_king_pos();
    let black_king = board.black_king_pos();

    if red_king.x != black_king.x {
        return false;
    }

    let min_y = red_king.y.min(black_king.y);
    let max_y = red_king.y.max(black_king.y);
    for y in (min_y + 1)..max_y {
        if board.piece_at(Position::new(red_king.x, y)).is_some() {
            return false;
        }
    }
    true
}

impl Clone for Board {
    fn clone(&self) -> Self {
        self.clone_for_search()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_test_board() -> Board {
        Board::new()
    }

    #[test]
    fn test_king_in_palace() {
        let board = make_test_board();
        assert!(is_legal_move(&board, Position::new(4, 9), Position::new(4, 8)));
        assert!(!is_legal_move(&board, Position::new(4, 9), Position::new(3, 9)));
        assert!(!is_legal_move(&board, Position::new(4, 9), Position::new(4, 7)));

        let board2 = Board::from_fen("4k4/9/9/9/9/4P3/9/9/9/3K5 w").unwrap();
        assert!(is_legal_move(&board2, Position::new(3, 9), Position::new(4, 9)));
        assert!(is_legal_move(&board2, Position::new(3, 9), Position::new(3, 8)));
        assert!(!is_legal_move(&board2, Position::new(3, 9), Position::new(2, 9)));
    }

    #[test]
    fn test_king_diagonal() {
        let board = make_test_board();
        assert!(!is_legal_move(&board, Position::new(4, 9), Position::new(5, 8)));
    }

    #[test]
    fn test_advisor_move() {
        let board = make_test_board();
        assert!(is_legal_move(&board, Position::new(3, 9), Position::new(4, 8)));
        assert!(!is_legal_move(&board, Position::new(3, 9), Position::new(3, 8)));
    }

    #[test]
    fn test_bishop_eye_blocked() {
        let board = Board::from_fen("4k4/9/9/9/9/9/9/9/1N6/2B1K4 w").unwrap();
        let bishop_from = Position::new(2, 9);
        let bishop_to = Position::new(0, 7);
        assert!(!is_legal_move(&board, bishop_from, bishop_to));
    }

    #[test]
    fn test_knight_leg_blocked() {
        let board = Board::from_fen("rnbakabnr/9/1c5c1/p1p1p1p1p/9/9/P1P1P1P1P/1C5C1/1p7/RNBAKABNR w").unwrap();
        assert!(!is_legal_move(&board, Position::new(1, 9), Position::new(2, 7)));
    }

    #[test]
    fn test_pawn_forward_only() {
        let board = make_test_board();
        assert!(is_legal_move(&board, Position::new(4, 6), Position::new(4, 5)));
        assert!(!is_legal_move(&board, Position::new(4, 6), Position::new(4, 7)));
        assert!(!is_legal_move(&board, Position::new(4, 6), Position::new(3, 6)));
    }

    #[test]
    fn test_cannon_capture() {
        let mut board = make_test_board();
        board.make_move_internal(Move::new(Position::new(1, 7), Position::new(4, 7)));
        board.make_move_internal(Move::new(Position::new(4, 3), Position::new(4, 5)));
        assert!(is_legal_move(&board, Position::new(4, 7), Position::new(4, 5)));
    }

    #[test]
    fn test_cannon_no_capture_over_piece() {
        let board = Board::from_fen("4k4/9/9/9/9/9/1p6/1C6/9/4K4 w").unwrap();
        let cannon_from = Position::new(1, 7);
        assert!(!is_legal_move(&board, cannon_from, Position::new(1, 5)));
    }

    #[test]
    fn test_rook_blocked() {
        let board = make_test_board();
        assert!(!is_legal_move(&board, Position::new(0, 9), Position::new(0, 5)));
    }

    #[test]
    fn test_check_detection() {
        let board = Board::from_fen("4k4/9/4p4/9/4C4/9/4P3/9/9/4K4 w").unwrap();
        assert!(is_in_check(&board, Side::Black));
    }

    #[test]
    fn test_no_check_initial() {
        let board = make_test_board();
        assert!(!is_in_check(&board, Side::Red));
        assert!(!is_in_check(&board, Side::Black));
    }

    #[test]
    fn test_flying_generals() {
        let board = Board::from_fen("4k4/9/9/9/9/4P4/9/9/9/4K4 w").unwrap();
        assert!(!is_legal_move(&board, Position::new(4, 4), Position::new(3, 4)));
        assert!(is_legal_move(&board, Position::new(4, 9), Position::new(3, 9)));
    }
}

File: xiangqi/engine/rules.py
from .types import Piece, PieceType, Side, Position, Move
from .board import Board
from typing import Optional


def is_legal_move(board: Board, from_pos: Position, to_pos: Position) -> bool:
    if not from_pos.is_valid() or not to_pos.is_valid():
        return False
    if from_pos == to_pos:
        return False

    piece = board.piece_at(from_pos)
    if piece is None or piece.side != board.side_to_move:
        return False

    target = board.piece_at(to_pos)
    if target is not None and target.side == piece.side:
        return False

    if not _is_piece_move_valid(board, piece, from_pos, to_pos):
        return False

    board.make_move_internal(Move(from_pos, to_pos))
    in_check = _is_in_check_fast(board, piece.side)
    board.undo_move_internal()

    return not in_check


def _is_piece_move_valid(board: Board, piece: Piece, from_pos: Position, to_pos: Position) -> bool:
    pt = piece.piece_type.value
    side = piece.side

    if pt == 0:  # King
        if side == Side.Red:
            if to_pos.x < 3 or to_pos.x > 5 or to_pos.y < 7 or to_pos.y > 9:
                return False
        else:
            if to_pos.x < 3 or to_pos.x > 5 or to_pos.y < 0 or to_pos.y > 2:
                return False
        dx = abs(to_pos.x - from_pos.x)
        dy = abs(to_pos.y - from_pos.y)
        return dx + dy == 1

    elif pt == 1:  # Advisor
        if side == Side.Red:
            if to_pos.x < 3 or to_pos.x > 5 or to_pos.y < 7 or to_pos.y > 9:
                return False
        else:
            if to_pos.x < 3 or to_pos.x > 5 or to_pos.y < 0 or to_pos.y > 2:
                return False
        dx = abs(to_pos.x - from_pos.x)
        dy = abs(to_pos.y - from_pos.y)
        return dx == 1 and dy == 1

    elif pt == 2:  # Bishop
        if side == Side.Red:
            if to_pos.y < 5:
                return False
        else:
            if to_pos.y > 4:
                return False
        dx = abs(to_pos.x - from_pos.x)
        dy = abs(to_pos.y - from_pos.y)
        if dx != 2 or dy != 2:
            return False
        eye_x = from_pos.x + (to_pos.x - from_pos.x) // 2
        eye_y = from_pos.y + (to_pos.y - from_pos.y) // 2
        return board.cells[eye_y * 9 + eye_x] is None

    elif pt == 3:  # Knight
        dx = to_pos.x - from_pos.x
        dy = to_pos.y - from_pos.y
        adx = abs(dx)
        ady = abs(dy)
        if not ((adx == 1 and ady == 2) or (adx == 2 and ady == 1)):
            return False
        if adx == 2:
            block_idx = from_pos.y * 9 + (from_pos.x + dx // 2)
        else:
            block_idx = (from_pos.y + dy // 2) * 9 + from_pos.x
        return board.cells[block_idx] is None

    elif pt == 4:  # Rook
        if from_pos.x != to_pos.x and from_pos.y != to_pos.y:
            return False
        return _is_path_clear_fast(board, from_pos, to_pos)

    elif pt == 5:  # Cannon
        if from_pos.x != to_pos.x and from_pos.y != to_pos.y:
            return False
        target = board.cells[to_pos.y * 9 + to_pos.x]
        if target is not None:
            return _count_between_fast(board, from_pos, to_pos) == 1
        else:
            return _is_path_clear_fast(board, from_pos, to_pos)

    elif pt == 6:  # Pawn
        dx = abs(to_pos.x - from_pos.x)
        dy = to_pos.y - from_pos.y
        forward = -1 if side == Side.Red else 1
        if side == Side.Red:
            past_river = from_pos.y <= 4
        else:
            past_river = from_pos.y >= 5
        if past_river:
            return (dy == forward and dx == 0) or (dy == 0 and dx == 1)
        else:
            return dy == forward and dx == 0

    return False


def _is_path_clear_fast(board: Board, from_pos: Position, to_pos: Position) -> bool:
    dx = to_pos.x - from_pos.x
    dy = to_pos.y - from_pos.y
    steps = max(abs(dx), abs(dy))
    if steps == 0:
        return False
    step_x = dx // steps
    step_y = dy // steps

    for i in range(1, steps):
        idx = (from_pos.y + step_y * i) * 9 + (from_pos.x + step_x * i)
        if board.cells[idx] is not None:
            return False
    return True


def _count_between_fast(board: Board, from_pos: Position, to_pos: Position) -> int:
    dx = to_pos.x - from_pos.x
    dy = to_pos.y - from_pos.y
    steps = max(abs(dx), abs(dy))
    step_x = dx // steps
    step_y = dy // steps

    count = 0
    for i in range(1, steps):
        idx = (from_pos.y + step_y * i) * 9 + (from_pos.x + step_x * i)
        if board.cells[idx] is not None:
            count += 1
    return count


def is_in_check(board: Board, side: Side) -> bool:
    return _is_in_check_fast(board, side)


def _is_in_check_fast(board: Board, side: Side) -> bool:
    if side == Side.Red:
        king_x, king_y = board.red_king_pos.x, board.red_king_pos.y
    else:
        king_x, king_y = board.black_king_pos.x, board.black_king_pos.y

    opponent = Side.Black if side == Side.Red else Side.Red
    king_idx = king_y * 9 + king_x

    # Check flying generals
    if board.red_king_pos.x == board.black_king_pos.x:
        ry = board.red_king_pos.y
        by = board.black_king_pos.y
        min_y = min(ry, by)
        max_y = max(ry, by)
        clear = True
        for y in range(min_y + 1, max_y):
            if board.cells[y * 9 + board.red_king_pos.x] is not None:
                clear = False
                break
        if clear:
            return True

    # Check all opponent pieces for attacks on king
    for y in range(10):
        for x in range(9):
            idx = y * 9 + x
            piece = board.cells[idx]
            if piece is None or piece.side != opponent:
                continue

            pt = piece.piece_type.value

            if pt == 4:  # Rook - same row or column, path clear
                if x == king_x or y == king_y:
                    if _is_path_clear_fast(board, Position(x, y), Position(king_x, king_y)):
                        return True

            elif pt == 5:  # Cannon - same row or column, exactly 1 between
                if x == king_x or y == king_y:
                    if _count_between_fast(board, Position(x, y), Position(king_x, king_y)) == 1:
                        return True

            elif pt == 3:  # Knight
                dx = abs(king_x - x)
                dy = abs(king_y - y)
                if (dx == 1 and dy == 2) or (dx == 2 and dy == 1):
                    if dx == 2:
                        block_idx = y * 9 + (x + (king_x - x) // 2)
                    else:
                        block_idx = (y + (king_y - y) // 2) * 9 + x
                    if board.cells[block_idx] is None:
                        return True

            elif pt == 6:  # Pawn
                dx = abs(king_x - x)
                dy = king_y - y
                forward = -1 if piece.side == Side.Red else 1
                if piece.side == Side.Red:
                    past_river = y <= 4
                else:
                    past_river = y >= 5
                if past_river:
                    if (dy == forward and dx == 0) or (dy == 0 and dx == 1):
                        return True
                else:
                    if dy == forward and dx == 0:
                        return True

    return False